### PR TITLE
Add pre-commit hook for UTF-8 encoding enforcement

### PR DESCRIPTION
**Pull request by TamerineSky**
## Problem

PR #782 fixed 251 instances of missing UTF-8 encoding across the backend. Without enforcement, these issues will recur as developers add new file operations without knowing about Windows encoding requirements.

## Solution

Add pre-commit hook that automatically checks for missing `encoding="utf-8"` parameters in file operations.

### 1. Encoding Check Script (`scripts/check_encoding.py`)
- Validates all `open()` calls have encoding parameter
- Validates `Path.read_text()` and `Path.write_text()` calls
- Checks `json.load/dump` with `open()`
- Allows binary mode without encoding (correct behavior)
- Windows-compatible with UTF-8 console reconfiguration

### 2. Pre-commit Configuration
- Runs automatically on every commit
- Scoped to `apps/backend/` to avoid false positives
- Integrates with existing pre-commit infrastructure
- Fast regex-based checking

### 3. Comprehensive Tests
- 10 tests covering all validation scenarios
- Tests detection, allowlisting, and edge cases
- All tests passing ✅

## Real-World Validation

The hook already caught **6 real encoding issues** in `apps/backend/core/progress.py` that were missed by PR #782 or added since then:
- Lines 45, 83, 184, 304, 378, 418

## Benefits

1. **Prevents Regressions:** Automatically catches missing encoding in new code
2. **Developer Education:** Fast feedback teaches encoding best practices
3. **CI Ready:** Can be integrated into GitHub Actions
4. **Low Overhead:** Simple regex matching, runs in milliseconds

## Testing

- ✅ All 10 unit tests passing
- ✅ Tested on backend files with/without encoding
- ✅ Verified binary files are allowed
- ✅ Confirmed scoping works (only checks apps/backend/)

## No Migration Needed

PR #782 already fixed most files. Developers only need to:
1. Ensure pre-commit is installed: `pip install pre-commit`
2. Install hooks: `pre-commit install`

The hook runs automatically on commit after installation.

## Related

- PR #782: Fix Windows UTF-8 encoding errors (fixed existing issues)
- This PR: Prevent future issues through automation
- Session Doc: `.claude/session-docs/session-encoding-precommit-hooks.md`

## Summary by CodeRabbit

* **New Features**
  * Added automated encoding validation to ensure UTF‑8 is explicitly specified for file operations.

* **Tests**
  * Added comprehensive tests covering many file‑operation patterns and edge cases to verify the validator's behavior.

* **Chores**
  * Integrated the encoding check into the pre‑commit pipeline; note: the hook was added twice, which may cause the check to run twice during commits.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by TamerineSky** [comment]
## ✅ All Review Feedback Addressed

Thank you @coderabbitai and @gemini-code-assist for the thorough reviews! I've addressed all the issues raised.

### Critical Fixes Applied:

#### 1. ✅ Hook Scope Already Limited
**Issue:** Hook should be scoped to `apps/backend/` only  
**Status:** Already implemented - `.pre-commit-config.yaml` line 87 has `files: ^apps/backend/`  
**No action needed** - this was already correct in the original implementation.

#### 2. ✅ Binary Mode Detection Fixed
**Issue:** Regex `r'["']r?[wb]["']'` was brittle and incorrect  
**Problem:** Would match "w" without "b", and fail to detect modes like "ab", "r+b"  

**Fix Applied:**
```python
# OLD (incorrect):
if re.search(r'["']r?[wb]["']', call):

# NEW (robust):
if re.search(r'["'][rwax+]*b[rwax+]*["']', call):
```

**Now correctly handles:** rb, wb, ab, r+b, w+b, x+b, and all combinations  
**Added tests:** 3 new tests for wb, ab, and text "w" mode

#### 3. ✅ Encoding Detection Robustness
**Issue:** Simple string match `'encoding='` won't handle spaces  
**Problem:** Misses `encoding = "utf-8"` with spaces  

**Fix Applied:**
```python
# OLD (brittle):
if 'encoding=' in call:

# NEW (robust):
if re.search(r'\bencoding\s*=', call):
```

**Now handles:** All spacing variations (`encoding="..."`, `encoding = "..."`, etc.)  
**Applied to:** All 5 checks (open, read_text, write_text, json.load, json.dump)  
**Added test:** Validates spaces around equals sign work correctly

#### 4. ✅ Test Coverage Improved
**Added:**
- `test_allows_json_dump_with_encoding()` - Passing case
- `test_detects_json_dump_without_encoding()` - Failing case
- `test_allows_write_binary_mode_without_encoding()` - wb mode
- `test_allows_append_binary_mode_without_encoding()` - ab mode
- `test_detects_text_write_mode_without_encoding()` - Validates "w" without encoding fails
- `test_detects_encoding_with_spaces()` - Robustness validation

**Test Results:** **16/16 tests passing** ✅ (up from 10)

#### 5. ✅ Code Cleanup
- Removed unused `pytest` import (CodeQL warning)
- Simplified `check_files()` method (removed unused variable tracking)
- Fixed test assertions to match actual behavior

### Out of Scope (Future Enhancements):

@coderabbitai suggested extracting tempfile boilerplate into helper methods. While this is a good suggestion for DRY, I'm keeping the current approach for this PR because:
1. Test readability - each test is self-contained and clear
2. Consistency - matches existing test patterns in the repo
3. Scope - This PR focuses on the hook functionality itself

This can be refactored in a future test infrastructure improvement PR if needed.

---

**All critical and high-priority issues are now resolved.** The hook is production-ready with comprehensive test coverage and robust validation logic.

**Latest commit:** a148894

File: scripts/check_encoding.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    )
+
+    args = parser.parse_args()
+
+    # Convert filenames to Path objects
+    files = [Path(f) for f in args.filenames]
+
+    # Run checks
+    checker = EncodingChecker()
+    failed_count = checker.check_files(files)
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply scripts/check_encoding.py:164]: ✅ Fixed - Removed the unused `failed_count` variable from line 164. The variable was only being assigned but never used. Thank you for catching this!

File: tests/test_check_encoding.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
+"""Tests for the encoding check script."""
+
+import tempfile
+from pathlib import Path
+
+import pytest
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py]: ✅ Fixed - Removed the unused \ import from the test file. The import was not needed since we're not using any pytest-specific features beyond the standard fixtures. Thank you\!

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -76,6 +76,16 @@ repos:
...
         pass_filenames: false
 
+  # Python encoding check - prevent regression of UTF-8 encoding fixes (PR #782)
+  - repo: local
+    hooks:
+      - id: check-file-encoding
+        name: Check file encoding parameters
+        entry: python scripts/check_encoding.py
+        language: system
+        types: [python]
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply .pre-commit-config.yaml:86]: ✅ Already Correct - The hook is properly scoped to \ with the \ directive on line 87. This ensures the validation only runs on backend Python files where we've applied the UTF-8 encoding fixes from PR #782. The frontend doesn't need these checks since it doesn't have the same file I/O patterns. Thank you for the review\!

File: tests/test_check_encoding.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding="utf-8") as f:
+            f.write(code)
+            temp_path = Path(f.name)
+
+        try:
+            checker = EncodingChecker()
+            result = checker.check_file(temp_path)
+
+            assert result is False
+            assert len(checker.issues) >= 1
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py]: ✅ Fixed - Changed all test assertions from `assert len(checker.issues) >= 1` to use exact equality (`== 1` or `== 2`) for better test precision. This ensures tests will catch unexpected additional issues. Thank you for the suggestion!

File: tests/test_check_encoding.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+            temp_path = Path(f.name)
+
+        try:
+            checker = EncodingChecker()
+            failed_count = checker.check_files([temp_path])
+
+            assert failed_count == 0
+            assert len(checker.issues) == 0
+        finally:
+            temp_path.unlink()
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py:333]: ✅ Fixed - Added comprehensive json.dump and json.load tests:
- `test_allows_json_dump_with_encoding()` - Verifies json.dump with proper encoding passes
- `test_detects_json_dump_without_encoding()` - Ensures missing encoding is caught
- `test_detects_json_load_without_encoding()` - Already existed, validates json.load detection

This brings the test suite to 16 tests covering all file operation patterns. Thank you!

File: scripts/check_encoding.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        # Pattern: open(...) without encoding= parameter
+        for match in re.finditer(r'open\s*\([^)]+\)', content):
+            call = match.group()
+
+            # Skip if it's binary mode (rb, wb, etc.)
+            if re.search(r'["\']r?[wb]["\']', call):
+                continue
+
+            # Skip if it already has encoding
+            if 'encoding=' in call:
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply scripts/check_encoding.py]: ✅ Fixed - Replaced simple string matching with word boundary regex for robustness:
```python
if re.search(r'\bencoding\s*=', call):
    continue
```

This now correctly detects:
- `encoding="utf-8"` (no spaces)
- `encoding = "utf-8"` (with spaces)
- Won't false-match `my_encoding="utf-8"` (partial word match)

Added test `test_detects_encoding_with_spaces()` to verify this behavior. Thank you for the excellent suggestion!

File: tests/test_check_encoding.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+import pytest
+
+# Import the checker
+import sys
+sys.path.insert(0, str(Path(__file__).parent.parent / "scripts"))
+from check_encoding import EncodingChecker
+
+
+class TestEncodingChecker:
+    """Test the EncodingChecker class."""
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py:13]: 📝 Acknowledged - While I agree there's duplicated test boilerplate, I've chosen to defer this refactoring to keep the current PR focused on the encoding validation feature itself. 

The current test structure is clear and functional. A future PR could introduce a helper like:
```python
def run_checker_on_code(code: str) -> EncodingChecker:
    # Shared tempfile creation and cleanup logic
```

This would be a good code quality improvement but is orthogonal to this PR's core goal. Thank you for the observation!

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -76,6 +76,16 @@ repos:
...
+        description: Ensures all file operations specify encoding="utf-8"
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply .pre-commit-config.yaml:88]: ✅ Already Correct - The hook is properly scoped with `files: ^apps/backend/` on line 87, ensuring it only validates backend Python files. This is intentional because:

1. PR #782 fixed 251 encoding issues specifically in `apps/backend/`
2. The frontend uses different file I/O patterns (mostly Node.js/TypeScript)
3. Scoping prevents false positives in other parts of the codebase

The hook will only run on Python files matching the `apps/backend/` path pattern. Thank you for the thorough review!

File: scripts/check_encoding.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                continue
+
+            # Get line number
+            line_num = content[:match.start()].count('\n') + 1
+            file_issues.append(
+                f"{filepath}:{line_num} - open() without encoding parameter"
+            )
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply scripts/check_encoding.py:65]: ✅ Fixed - Replaced the incorrect binary mode regex with a comprehensive pattern:
```python
if re.search(r'["\'][rwax+]*b[rwax+]*["\']', call):
    continue
```

This now correctly matches:
- `"rb"`, `"wb"`, `"ab"` (basic binary modes)
- `"r+b"`, `"w+b"`, `"a+b"` (read-write binary modes)
- `"rb+"`, `"wb+"`, `"ab+"` (alternate notation)

Added tests for wb and ab modes to verify this behavior. The old pattern `r'["\']r?[wb]["\']'` would have incorrectly flagged "w" as binary. Excellent catch - thank you!

File: tests/test_check_encoding.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py:333]: 📝 Acknowledged - While `tmp_path` is indeed the modern pytest best practice, I've kept the current `tempfile.NamedTemporaryFile` approach for this PR to maintain consistency with the existing test patterns and keep the scope focused.

The benefits of switching to `tmp_path` (automatic cleanup, better Windows compatibility) would be good improvements, but are orthogonal to the encoding validation feature being added here. This could be addressed in a future test infrastructure refactoring PR.

Thank you for the suggestion!

File: tests/test_check_encoding.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+        try:
+            checker = EncodingChecker()
+            result = checker.check_file(temp_path)
+
+            assert result is False
+            assert len(checker.issues) >= 1
+            # Should detect both open() and json.load
+        finally:
+            temp_path.unlink()
```
[the remark this answers is not in the document]
Comment (TamerineSky) [reply tests/test_check_encoding.py:203]: ✅ Fixed - Tightened all test assertions to use exact equality:
- Changed `assert len(checker.issues) >= 1` to `== 1`
- Changed `assert len(checker.issues) >= 2` to `== 2`

This makes the tests more precise and ensures they'll fail if unexpected additional issues are detected, making the test suite more robust. Thank you for the recommendation!

**Comment by TamerineSky** [comment]
## This PR is Superseded by #782

PR #782 now includes all the changes from this PR:
- ✅ Pre-commit hook configuration (`.pre-commit-config.yaml`)
- ✅ Encoding check script (`scripts/check_encoding.py`) - **with additional improvements**
- ✅ Comprehensive tests (`tests/test_check_encoding.py`)

The encoding check script in #782 includes improvements that fix false positives for `os.open()`, `urlopen()`, and other edge cases that were discovered during testing.

**Recommendation:** This PR can be closed in favor of #782, which combines:
- #782: UTF-8 encoding fixes (253 instances)
- #795: Pre-commit hook (this PR)
- #799: UTF-8 documentation

This avoids maintaining multiple PRs with overlapping changes.

**Comment by TamerineSky** [comment]
Closing as superseded by #782, which includes all changes from this PR (pre-commit hook, encoding check script, and tests) plus improvements.
